core/src/main/java/io/micronaut/core/convert/MutableConversionService.java

```diff
@@ -0,0 +1,63 @@
+/*
+ * Copyright 2017-2020 original authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.micronaut.core.convert;
+
+import io.micronaut.core.annotation.NonNull;
+
+import java.util.function.Function;
+
+/**
+ * A version of {@link ConversionService} that supports adding new converters.
+ *
+ * @author Denis Stepanov
+ * @since 4.0.0
+ */
+public interface MutableConversionService extends ConversionService {
+
+    /**
+     * Creates a new mutable conversion service that extends the shared conversion service.
+     * In most cases the mutable service from the bean context should be used.
+     *
+     * @return A new mutable conversion service.
+     */
+    @NonNull
+    static MutableConversionService create() {
+        return new DefaultMutableConversionService();
+    }
+
+    /**
+     * Adds a type converter.
+     *
+     * @param sourceType    The source type
+     * @param targetType    The target type
+     * @param typeConverter The type converter
+     * @param <S>           The source generic type
+     * @param <T>           The target generic type
+     */
+    <S, T> void addConverter(@NonNull Class<S> sourceType, @NonNull Class<T> targetType, @NonNull Function<S, T> typeConverter);
```

Comment:
Return `MutableConversionService` to make the API fluent

Reply:
Do you want to have it generic like it was before?

Reply:
Makes sense

Reply:
I'm looking at it again, and I don't like to have it in this way. There are zero usages of chaining in the current codebase. The registration API defines this signature: `void register(ConversionService<?> conversionService)` which prevents to do the chaining. IMHO the generic parameter only makes the code ugly, and the chaining is not very important in this case. I would prefer not to have it. WDYT @micronaut-projects/core-developers ?

Reply:
It is probably fine without it but the generic argument remains a potentially large source of breaking changes